File: src/plnflr/render/svg.py
"""Server-side SVG of the whole install."""

from __future__ import annotations

from plnflr.domain.models import LayoutPlan, Ring
from plnflr.engine.expansion import bbox


def _path(ring: Ring) -> str:
    verts = ring.vertices
    if not verts:
        return ""
    start = verts[0]
    parts = [f"M {start.x_mm} {start.y_mm}"]
    for v in verts[1:]:
        parts.append(f"L {v.x_mm} {v.y_mm}")
    parts.append("Z")
    return " ".join(parts)


def _poly_path(points: tuple) -> str:
    if not points:
        return ""
    start = points[0]
    parts = [f"M {start.x_mm} {start.y_mm}"]
    for v in points[1:]:
        parts.append(f"L {v.x_mm} {v.y_mm}")
    parts.append("Z")
    return " ".join(parts)


def plan_to_svg(plan: LayoutPlan, *, max_px: int = 900) -> str:
    min_x, min_y, max_x, max_y = bbox(plan.room)
    width = max(max_x - min_x, 1)
    height = max(max_y - min_y, 1)
    evenodd = _path(plan.room.outer)
    for hole in plan.room.holes:
        evenodd += " " + _path(hole)
    gap_path = _path(plan.room.outer) + " " + _path(plan.inset.outer)
    for hole in plan.room.holes:
        gap_path += " " + _path(hole)
    for hole in plan.inset.holes:
        gap_path += " " + _path(hole)
    pieces = []
    for piece in plan.pieces:
        cls = "pln-row-even" if piece.row_index % 2 == 0 else "pln-row-odd"
        if piece.kind in {"clip", "rip", "tile_cut"}:
            cls += " pln-clip"
        pieces.append(
            f'<path class="{cls}" data-piece-id="{piece.piece_id}" '
            f'd="{_poly_path(piece.geometry)}" />'
        )
    labels = []
    seen: set[int] = set()
    for piece in plan.pieces:
        if piece.row_index in seen:
            continue
        seen.add(piece.row_index)
        xs = [v.x_mm for v in piece.geometry]
        ys = [v.y_mm for v in piece.geometry]
        cx = (min(xs) + max(xs)) // 2
        cy = (min(ys) + max(ys)) // 2
        labels.append(
            f'<text class="pln-label" x="{cx}" y="{cy}" text-anchor="middle" '
            f'dominant-baseline="middle">{piece.row_index + 1}</text>'
        )
    body = "\n".join(
        [
            f'<path class="pln-room" fill-rule="evenodd" d="{evenodd}" />',
            f'<path class="pln-gap" fill-rule="evenodd" d="{gap_path}" />',
            *pieces,
            *labels,
        ]
    )
    return (
        f'<svg class="pln-preview" xmlns="http://www.w3.org/2000/svg" '
        f'viewBox="{min_x} {min_y} {width} {height}" '
        f'width="{max_px}" role="img" aria-label="Plan ułożenia podłogi">'
        f"{body}</svg>"
    )

File: Kernel/Sources/PlnFlrLayout/SVG.swift
public func planToSvg(_ plan: LayoutPlan, maxPx: Int = 900) -> String {
    let bounds = bbox(plan.room)
    let width = max(bounds.maxX - bounds.minX, 1)
    let height = max(bounds.maxY - bounds.minY, 1)
    var evenodd = path(plan.room.outer)
    for hole in plan.room.holes { evenodd += " " + path(hole) }
    var gapPath = path(plan.room.outer) + " " + path(plan.inset.outer)
    for hole in plan.room.holes { gapPath += " " + path(hole) }
    for hole in plan.inset.holes { gapPath += " " + path(hole) }
    var pieces: [String] = []
    for piece in plan.pieces {
        let parity = piece.rowIndex % 2 == 0 ? "pln-row-even" : "pln-row-odd"
        var cls = "\(parity) pln-zone-\(piece.zoneIndex)"
        if piece.kind == .clip || piece.kind == .rip || piece.kind == .tileCut {
            cls += " pln-clip"
        }
        pieces.append(
            "<path class=\"\(cls)\" data-piece-id=\"\(piece.pieceId)\" data-zone=\"\(piece.zoneIndex)\" d=\"\(polyPath(piece.geometry))\" />"
        )
    }
    var labels: [String] = []
    var seen: Set<String> = []
    for piece in plan.pieces {
        let key = "\(piece.zoneIndex)-\(piece.rowIndex)"
        if seen.contains(key) { continue }
        seen.insert(key)
        let xs = piece.geometry.map(\.xMm)
        let ys = piece.geometry.map(\.yMm)
        let cx = ((xs.min() ?? 0) + (xs.max() ?? 0)) / 2
        let cy = ((ys.min() ?? 0) + (ys.max() ?? 0)) / 2
        labels.append(
            "<text class=\"pln-label\" x=\"\(cx)\" y=\"\(cy)\" text-anchor=\"middle\" dominant-baseline=\"middle\">\(piece.rowIndex + 1)</text>"
        )
    }
    var extras: [String] = []
    if let divider = plan.divider {
        extras.append(
            "<line class=\"pln-divider\" x1=\"\(divider.0.xMm)\" y1=\"\(divider.0.yMm)\" x2=\"\(divider.1.xMm)\" y2=\"\(divider.1.yMm)\" />"
        )
    }
    for strip in plan.thresholds {
        extras.append("<path class=\"pln-threshold\" d=\"\(path(strip.geometry))\" />")
    }
    for window in plan.windows {
        extras.append(
            "<line class=\"pln-window\" x1=\"\(window.start.xMm)\" y1=\"\(window.start.yMm)\" x2=\"\(window.end.xMm)\" y2=\"\(window.end.yMm)\" />"
        )
    }
    let body = [
        "<path class=\"pln-room\" fill-rule=\"evenodd\" d=\"\(evenodd)\" />",
        "<path class=\"pln-gap\" fill-rule=\"evenodd\" d=\"\(gapPath)\" />",
    ] + pieces + extras + labels
    return
        "<svg class=\"pln-preview\" xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"\(bounds.minX) \(bounds.minY) \(width) \(height)\" width=\"\(maxPx)\" role=\"img\" aria-label=\"Plan ułożenia podłogi\">\(body.joined())</svg>"
}

private func path(_ ring: Ring) -> String {
    polyPath(ring.vertices)
}

private func polyPath(_ points: [Vertex]) -> String {
    guard let start = points.first else { return "" }
    var parts = ["M \(start.xMm) \(start.yMm)"]
    for v in points.dropFirst() {
        parts.append("L \(v.xMm) \(v.yMm)")
    }
    parts.append("Z")
    return parts.joined(separator: " ")
}
